service-base/src/main/java/org/eclipse/hono/service/auth/ExternalJwtAuthTokenValidator.java

```diff
@@ -0,0 +1,200 @@
+import org.eclipse.hono.util.RegistryManagementConstants;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.JwsHeader;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.MalformedJwtException;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.SigningKeyResolverAdapter;
+import io.jsonwebtoken.UnsupportedJwtException;
+import io.jsonwebtoken.security.SignatureException;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A class to validate a JSON Web Token (JWT) against a CredentialsObject containing a public key/certificate.
+ */
+public class ExternalJwtAuthTokenValidator implements AuthTokenValidator {
+
+    static final int ALLOWED_CLOCK_SKEW = 600;
+    private static final String EXPECTED_TOKEN_TYPE = "JWT";
+    private CredentialsObject credentialsObject;
+
+    public CredentialsObject getCredentialsObject() {
+        return credentialsObject;
+    }
+
+    public void setCredentialsObject(final CredentialsObject credentialsObject) {
+        this.credentialsObject = credentialsObject;
+    }
+
+    @Override
+    public Jws<Claims> expand(final String token) {
+        Objects.requireNonNull(token);
+        final var builder = Jwts.parserBuilder()
+                .setSigningKeyResolver(new SigningKeyResolverAdapter() {
+
+                    @Override
+                    public Key resolveSigningKey(final JwsHeader header, final Claims claims) {
+
+                        final var tokenType = Optional.ofNullable(header.getType())
+                                .orElseThrow(
+                                        () -> new SignatureException("token does not contain required typ header."));
+                        if (!tokenType.equalsIgnoreCase(EXPECTED_TOKEN_TYPE)) {
+                            throw new SignatureException(String.format(
+                                    "typ field in token header is invalid. Must be \"%s\".", EXPECTED_TOKEN_TYPE));
+                        }
+
+                        final var algorithm = Optional.ofNullable(header.getAlgorithm())
+                                .orElseThrow(
+                                        () -> new SignatureException("token does not contain required alg header."));
+                        final SignatureAlgorithm signatureAlgorithm = SignatureAlgorithm.forName(algorithm);
+
+                        checkValidityOfExpirationAndCreationTime(claims.getExpiration(), claims.getIssuedAt());
+                        claims.setNotBefore(claims.getIssuedAt());
+
+                        final List<JsonObject> secrets = getCredentialsObject().getCandidateSecrets();
+                        final Optional<JsonObject> optionalSecret = secrets.stream()
+                                .filter(secret -> signatureAlgorithm.getFamilyName()
+                                        .startsWith(secret.getString(JwsHeader.ALGORITHM)))
+                                .findFirst();
```

Comment:
what if there are multiple public keys registered for a device, i.e. when rotating keys? This implementation will always only try to verify the signature using the first key, won't it?

Reply:
yes, but secrets.stream() is a stream of only valid secrets (see line before), so after the filter there should usually only be one key but even if there are more it should not matter as they all are valid and of the right type. Does this make sense?

Reply:
IMHO the use case for registering multiple public keys for a device in Hono is key rotation, i.e. you want to roll out a new set of private/public keys to a device and want to make sure that for a period of time the device will be able to authenticate using either the old private key or the new private key. This allows for the rollout of the new key pair to be done anytime during this period. For this to work, you would add the new public key to the device's secrets and define its *not-before* and *not-after* properties to overlap with the existing public key's validity period. However, for this to work, the protocol adapter would need to consider both public keys when trying to validate the JWT's signature.

Reply:
Ah I see, you are absolutely right. I made a change so that all valid keys are being considered. This enables the possibility to use rotating keys.